automated/linux/mmtests/mmtests.sh

```diff
@@ -12,199 +9,242 @@ OUTPUT="$(pwd)/output"
+    benchmarks=()
+    while IFS= read -r benchmark; do
+      benchmarks+=("$benchmark")
+    done < <(find "${log_dir}/iter-0" -type d | grep -E 'logs$' | grep -Eo 'iter-0/.+/logs' | cut -d '/' -f 2)
+    # Iterate through found benchmark names to extract results
+    for benchmark in "${benchmarks[@]}"; do
+      # Build JSON file name
+      # MARKERwords added intentionally, this allows to parse filename
+      # easily in the future.
+      results_json=BENCHMARK${benchmark}_CONFIG${testname}.json
+      # Call parser
+      ${MMTEST_EXTR} -d ${results_loc} -b "${benchmark}" -n "${full_testname}" --print-json > "${results_json}"
+      # Add JSON file name to array
+      jsons+=("${results_json}")
+    done
+  done
+  # It's required to return of array separated by new line
+  printf "%s\n" "${jsons[@]}"
+}
+
+collect_details() {
+  # Collect benchmark run details
+  MEMTOTAL_BYTES=$(free -b | grep Mem: | awk '{print $2}')
+  NUMCPUS=$(grep -c '^processor' /proc/cpuinfo)
+  NUMNODES=$(grep ^Node /proc/zoneinfo | awk '{print $2}' | sort | uniq | wc -l)
+  LLC_INDEX=$(find /sys/devices/system/cpu/ -type d -name "index*" | sed -e 's/.*index//' | sort -n | tail -1)
+  NUMLLCS=$(grep . /sys/devices/system/cpu/cpu*/cache/index"$LLC_INDEX"/shared_cpu_map | awk -F : '{print $NF}' | sort | uniq | wc -l)
+  KERNEL_VERSION=$(uname -r)
+  cat <<EOF
+{
+  "MEMTOTAL_BYTES": "${MEMTOTAL_BYTES:-}",
+  "NUMCPUS": "${NUMCPUS:-}",
+  "NUMNODES": "${NUMNODES:-}",
+  "LLC_INDEX": "${LLC_INDEX:-}",
+  "NUMLLCS": "${NUMLLCS:-}",
+  "KERNEL_VERSION": "${KERNEL_VERSION:-}",
+  "MMTESTS_ITERATIONS": "${MMTESTS_ITERATIONS:-}",
+  "MMTESTS_CONFIG_FILE": "${MMTESTS_CONFIG_FILE:-}"
+}
+EOF
+}
+
+collect_results() {
+  # Collect benchmark run details in JSON object.
+  details=$(collect_details)
+  # Dump details to temp file
+  details_file=$(mktemp)
+  echo "$details" > "$details_file"
+  # Extract results data from available logs for each benchmark in JSON format.
+  mapfile -t jsons < <(extract_json)
+  for json in "${jsons[@]}"; do
+    # Create a temp file to hold the merged JSON
+    merge_file=$(mktemp)
+    # Merge details and results JSON
+    jq -n \
+        --argfile d "$details_file" \
+        --argfile r "$json" \
+        '{details: $d, results: $r}' > "$merge_file"
+    # Replace results file
+    mv "$merge_file" "$json"
+  done
 }
 
 ! check_root && error_msg "Please run this script as root."
 
-# Test installation.
-if [ "${SKIP_INSTALL}" = "true" ] || [ "${SKIP_INSTALL}" = "True" ]; then
-	info_msg "${MMTESTS_TYPE_NAME} installation skipped"
+if [ "${SKIP_INSTALL}" = "true" ]; then
+  info_msg "Installation skipped"
 else
-	install
+  # Install system-wide dependencies.
+  install_system_deps
+  # Install benchmark and Perl dependencies.
+  prepare_system
+  # Clone MMTests repository.
+  get_test_program "${TEST_GIT_URL}" "${TEST_DIR}" "${TEST_PROG_VERSION}" "${TEST_PROGRAM}"
+  create_out_dir "${OUTPUT}"
```

Comment:
This refactoring wont work.
1. prepare system can't be done before get_test_program if that's what you specify.
2. you might don't want to install any build dependencies to the rootfs, but you might want to clone the repository and build mmtests test suites.

Reply:
1. Correct.
2. Maybe, but in this case extra steps for deps installation will be performed.